crypto/thread/task.c

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2019-2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/configuration.h>
+#include <openssl/crypto.h>
+
+#if defined(OPENSSL_THREADS)
+
+# include "task.h"
+
+static void crypto_task_free(CRYPTO_TASK *t);
+
+int task_list_cmp(struct list *l, void *data)
+{
+    return container_of(l, struct crypto_task_st, list)->id == *(uint32_t *)data;
+}
+
+CRYPTO_TASK *crypto_task_new(CRYPTO_THREAD_ROUTINE start, void *data)
+{
+    CRYPTO_TASK *t;
+
+    t = OPENSSL_zalloc(sizeof(*t));
+    if (t == NULL)
+        return NULL;
+
+    t->state = CRYPTO_TASK_NO_STATE;
+
+    t->lock = CRYPTO_MUTEX_create();
+    t->cond = CRYPTO_CONDVAR_create();
+
+    if (t->lock == NULL || t->cond == NULL)
+        goto fail;
+
+    if (CRYPTO_MUTEX_init(t->lock) == 0)
+        goto fail;
+
+    if (CRYPTO_CONDVAR_init(t->cond) == 0)
+        goto fail;
+
+    t->data = data;
+    t->routine = start;
+    return t;
+
+fail:
+    crypto_task_free(t);
+    return NULL;
+}
+
+int crypto_task_clean(OSSL_LIB_CTX *ctx, void *vtask)
+{
+    CRYPTO_TASK *task;
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    task = vtask;
+    tdata = OSSL_LIB_CTX_GET_THREADS(ctx);
+
+    if (tdata == NULL || vtask == NULL)
+        return 0;
+
+    if (task->state != CRYPTO_TASK_FINISHED)
+        return 1;
+
+    CRYPTO_MUTEX_lock(tdata->lock);
+    list_del(&task->list);
+    CRYPTO_MUTEX_unlock(tdata->lock);
+
+    crypto_task_free(task);
+
+    return 1;
+}
+
+static void crypto_task_clean_list(OSSL_LIB_CTX *ctx, struct list *list)
+{
+    CRYPTO_TASK *task;
+    struct list *iter, *tmp;
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    tdata = OSSL_LIB_CTX_GET_THREADS(ctx);
+    if (tdata == NULL)
+        return;
+
+    CRYPTO_MUTEX_lock(tdata->lock);
+    list_for_each_safe(iter, tmp, list) {
+        task = container_of(iter, struct crypto_task_st, list);
+        if (task == NULL)
+            continue;
+        list_del(&task->list);
+        crypto_task_free(task);
+    }
+    CRYPTO_MUTEX_unlock(tdata->lock);
+}
+
+int crypto_task_clean_all(OSSL_LIB_CTX *ctx)
+{
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    tdata = OSSL_LIB_CTX_GET_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    OPENSSL_assert(list_size(&tdata->tasks.active) == 0);
```

Comment:
Use `ossl_assert`. We try to avoid `OPENSSL_assert` inside the library these days.